### PR TITLE
Draft market governance

### PR DESCRIPTION
**Pull request by eswak**
to start discussion and commenting on code

### CODE REVIEW

File: contracts/vcon/MarketGovernanceRewards.sol [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        } else {
+            uint256 slashing = (uint256(-profitAttributableToUser) *
+                vconPerEarnedUSD) / 1e18;
+            uint256 userBalance = IERC20(_vcon).balanceOf(user);
+            if (slashing > userBalance) {
+                slashing = userBalance;
+            }
+            IERC20(_vcon).transferFrom(user, address(this), slashing);
+        }
+    }
```
Comment (eswak) [inline contracts/vcon/MarketGovernanceRewards.sol:219]: I'm really unhappy with this part of the code

File: contracts/vcon/ERC20Lockable.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "solmate/tokens/ERC20.sol";
```
Comment (ElliotFriedman) [inline contracts/vcon/ERC20Lockable.sol:5]: not seeing the solmate library imported here

File: contracts/vcon/ERC20Lockable.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    /// @notice lock msg.sender's tokens on a locker
+    function lock(address locker) external {
+        _setLockAddress(msg.sender, locker);
+    }
+
+    /// @notice unlock a user that is locking on msg.sender
+    function unlock(address user) external {
+        if (msg.sender != lockAddress[user]) revert LockError(); // invalid unlocker
+        _unsetLockAddress(user);
+    }
```
Comment (ElliotFriedman) [inline contracts/vcon/ERC20Lockable.sol:64]: we need another feature that stops users from transferring tokens if the global transfer ability is paused

File: contracts/vcon/MarketGovernance.sol [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                IPCVDepositV2(movement.venueTo).harvest();
+                IPCVDepositV2(movement.venueTo).accrue();
+            }
+        }
+
+        {
+            // Refresh yields and indexes
+            // TODO: check if this "manual update list" is useful
+            for (
+                uint256 i = venuesToRefresh.length;
```
Comment (ElliotFriedman) [inline contracts/vcon/MarketGovernance.sol:103]: why do we need this set of venues to refresh?

File: contracts/vcon/MarketGovernance.sol [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+            // check allocations with 0.5% diff tolerance
+            for (uint256 i = gauges.length; i < gauges.length; i++) {
+                uint256 gaugeAllocation = ERC20Gauges(_vcon)
+                    .calculateGaugeAllocation(gauges[i], totalManagedValueUSD);
+                uint256 minBalanceUSD = (gaugeAllocation * 9950) / 10000;
+                require(
+                    balancesUSD[i] > minBalanceUSD,
+                    "MarketGovernance: insufficient allocation"
+                );
```
Comment (ElliotFriedman) [inline contracts/vcon/MarketGovernance.sol:144]: is this checking that the optimistic rebalance correctly occurred?

File: contracts/vcon/ERC20Lockable.sol [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    mapping(address => address) public lockAddress;
+
+    /// @dev Emitted when a `user` lock their tokens on `lock`.
+    event Lock(address indexed user, address indexed lock);
+
+    /// @dev Emitted when a `user` unlock their tokens from `lock`.
+    event Unlock(address indexed user, address indexed lock);
+
+    /// @notice set the locking address of an account
+    function _setLockAddress(address user, address userLock) internal {
```
Comment (ElliotFriedman) [inline contracts/vcon/ERC20Lockable.sol:28]: I'd prefer a locking mechanism that allows users to lock tokens up into a gauge, as well as global transfer restrictions that ensure you can only transfer tokens to whitelisted addresses as well as a global lock on all transfers.

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: contracts/vcon/ERC20Lockable.sol (hunk 2 above)
Comment (eswak) [reply contracts/vcon/ERC20Lockable.sol:5]: it's node module & I added an entry in remappings.txt, like we do for OZ library

File: contracts/vcon/ERC20Lockable.sol (hunk 3 above)
Comment (eswak) [reply contracts/vcon/ERC20Lockable.sol:64]: yep I've put this logic in VconV2 directly

File: contracts/vcon/MarketGovernance.sol (hunk 5 above)
Comment (eswak) [reply contracts/vcon/MarketGovernance.sol:144]: yes the call is permissionless, so anyone can specify the list of PCV movements to do at the end of the week, but this loop checks that the post-movement balances correspond to what has been voted on in the gauge system

**Pull request closed without merging**